src/com/osudroid/ui/v2/BeatmapInfoLayout.kt

```diff
@@ -0,0 +1,260 @@
+                }
+            }
+
+            linearContainer {
+                orientation = Orientation.Horizontal
+                spacing = 2f
+
+                arText = labeledBadge {
+                    label = "AR"
+                    value = "00.0"
+                    sizeVariant = SizeVariant.Small
+                }
+                odText = labeledBadge {
+                    label = "OD"
+                    value = "00.0"
+                    sizeVariant = SizeVariant.Small
+                }
+                csText = labeledBadge {
+                    label = "CS"
+                    value = "00.0"
+                    sizeVariant = SizeVariant.Small
+                }
+                hpText = labeledBadge {
+                    label = "HP"
+                    value = "00.0"
+                    sizeVariant = SizeVariant.Small
+                }
+            }
+        }
+    }
+
+
+    fun setBeatmapInfo(beatmapInfo: BeatmapInfo?) {
+        titleText.text = beatmapInfo?.titleText ?: "No selected beatmap"
+        artistText.text = beatmapInfo?.artistText ?: "Unknown"
+        versionText.text = beatmapInfo?.version ?: "Unknown"
+        setDifficultyStatistics(beatmapInfo)
+    }
+
+    fun setBeatmapInfo(roomBeatmap: RoomBeatmap?) {
+        titleText.text = roomBeatmap?.title ?: "No selected beatmap"
+        artistText.text = roomBeatmap?.artist ?: "Unknown"
+        versionText.text = roomBeatmap?.version ?: "Unknown"
+        setDifficultyStatistics(null)
+    }
+
+    /**
+     * Change the displayed difficulty statistics.
+     */
+    fun setDifficultyStatistics(beatmapInfo: BeatmapInfo?) {
+
+        circlesBadge.value = beatmapInfo?.hitCircleCount?.toString() ?: "0"
+        slidersBadge.value = beatmapInfo?.sliderCount?.toString() ?: "0"
+        spinnersBadge.value = beatmapInfo?.spinnerCount?.toString() ?: "0"
+
+        if (beatmapInfo == null) {
+            arText.value = "00.0"
+            odText.value = "00.0"
+            csText.value = "00.0"
+            hpText.value = "00.0"
+            starsText.text = "0.00"
+            bpmText.text = "0"
+            lengthText.text = "00:00"
+            return
+        }
+
+        val mods = ModMenu.enabledMods
+        val isPreciseMod = ModPrecise::class in mods
+        val totalSpeedMultiplier = calculateRateWithMods(mods.values, Double.POSITIVE_INFINITY)
+
+        val difficulty = beatmapInfo.getBeatmapDifficulty()
+
+        applyModsToBeatmapDifficulty(difficulty, GameMode.Droid, mods.values, true)
+
+        if (isPreciseMod) {
+            // Special case for OD. The Precise mod changes the hit window and not the OD itself, but we must
+            // map the hit window back to the original hit window for the user to understand the difficulty
+            // increase of the mod.
+            val greatWindow = PreciseDroidHitWindow(difficulty.od).greatWindow
+            difficulty.od = hitWindow300ToOverallDifficulty(greatWindow)
+        }
```

Comment:
This is not valid anymore as of https://github.com/osudroid/osu-droid/commit/bb6e066075de20976685e50c5af860bda1fd2268.

Reply:
Should I remove it ?

Reply:
Fixed [b5f55c9](https://github.com/osudroid/osu-droid/pull/572/commits/b5f55c9f000c5a55db1dc40158de41d24b180db6)